File: example/src/main/java/com/github/vivchar/example/pages/simple/PayloadFragment.java
package com.github.vivchar.example.pages.simple;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.github.vivchar.example.BaseScreenFragment;
import com.github.vivchar.example.R;
import com.github.vivchar.example.widgets.ItemOffsetDecoration;
import com.github.vivchar.rendererrecyclerviewadapter.DefaultDiffCallback;
import com.github.vivchar.rendererrecyclerviewadapter.RendererRecyclerViewAdapter;
import com.github.vivchar.rendererrecyclerviewadapter.ViewModel;
import com.github.vivchar.rendererrecyclerviewadapter.binder.ViewBinder;
import com.github.vivchar.rendererrecyclerviewadapter.binder.ViewProvider;

/**
 * Created by Vivchar Vitaly on 12/29/17.
 */

public class PayloadFragment extends BaseScreenFragment {

	public static final Integer TEXT_CHANGED = 1;
	public static final Integer DESCRIPTION_CHANGED = 2;

	private YourDataProvider mYourDataProvider = new YourDataProvider();
	private RendererRecyclerViewAdapter mAdapter;
	private RecyclerView mRecyclerView;

	@Nullable
	@Override
	public View onCreateView(final LayoutInflater inflater,
	                         @Nullable final ViewGroup container,
	                         @Nullable final Bundle savedInstanceState) {

		final View view = inflater.inflate(R.layout.fragment_list, container, false);

		mAdapter = new RendererRecyclerViewAdapter(getContext());

		mAdapter.setDiffCallback(new PayloadDiffCallback());

		mAdapter.registerRenderer(new ViewBinder<>(
				R.layout.item_payload_square, PayloadViewModel.class,
				(model, finder, payloads) -> {
					finder.setOnClickListener(R.id.text, v -> changeItem(model));
					final TextView textView = finder.find(R.id.text);

					if (payloads.isEmpty()) {
						/* full bind */
						textView.setText(model.getText());
						finder.setText(R.id.desciption, model.getDescription());
					} else {
						/* partially bind */
						final Object payload = payloads.get(0);
						if (payload.equals(TEXT_CHANGED)) {
							textView.setRotation(0);
							textView.animate().rotation(360).start();
							textView.setText(model.getText());
						} else if (payload.equals(DESCRIPTION_CHANGED)) {
							finder.setText(R.id.desciption, model.getDescription());
						}
					}
				}
		));
//		adapter.registerRenderer(...);
//		adapter.registerRenderer(...);

		mAdapter.setItems(mYourDataProvider.getPayloadItems());

		mRecyclerView = (RecyclerView) view.findViewById(R.id.recycler_view);
		mRecyclerView.setAdapter(mAdapter);
		mRecyclerView.setLayoutManager(new GridLayoutManager(getContext(), 3));
		mRecyclerView.addItemDecoration(new ItemOffsetDecoration(10));

		return view;
	}

	private void changeItem(@NonNull final PayloadViewModel model) {
		mAdapter.setItems(mYourDataProvider.getChangedPayloadItems(model));
	}

	private class PayloadDiffCallback extends DefaultDiffCallback<PayloadViewModel> {

		@Override
		public boolean areItemsTheSame(@NonNull final PayloadViewModel oldItem, @NonNull final PayloadViewModel newItem) {
			return oldItem.getID() == newItem.getID();
		}

		@Nullable
		@Override
		public Object getChangePayload(@NonNull final PayloadViewModel oldItem, @NonNull final PayloadViewModel newItem) {
			if (!oldItem.getText().equals(newItem.getText())) {
				return TEXT_CHANGED;
			}
			if (!oldItem.getDescription().equals(newItem.getDescription())) {
				return DESCRIPTION_CHANGED;
			}
			return super.getChangePayload(oldItem, newItem);
		}
	}

	public static class PayloadViewModel implements ViewModel {

		private final int mID;
		private final String mText;
		private final String mDescription;

		public PayloadViewModel(final int ID, final String text, final String description) {
			mID = ID;
			mText = text;
			mDescription = description;
		}

		public String getText() {
			return mText;
		}

		public int getID() {
			return mID;
		}

		@Override
		public boolean equals(final Object o) {
			if (this == o) {
				return true;
			}
			if (o == null || getClass() != o.getClass()) {
				return false;
			}

			final PayloadViewModel that = (PayloadViewModel) o;

			if (mID != that.mID) {
				return false;
			}
			return mText != null ? mText.equals(that.mText) : that.mText == null;
		}

		@Override
		public int hashCode() {
			int result = mID;
			result = 31 * result + (mText != null ? mText.hashCode() : 0);
			return result;
		}

		public String getDescription() {
			return mDescription;
		}
	}
}

File: example/src/main/java/com/github/vivchar/example/pages/simple/PayloadFragment.kt
package com.github.vivchar.example.pages.simple

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.GridLayoutManager
import com.github.vivchar.example.R
import com.github.vivchar.example.base.BaseFragment
import com.github.vivchar.example.databinding.FragmentListBinding
import com.github.vivchar.example.widgets.ItemOffsetDecoration
import com.github.vivchar.example.widgets.MyAdapter
import com.github.vivchar.rendererrecyclerviewadapter.DefaultDiffCallback
import com.github.vivchar.rendererrecyclerviewadapter.ViewFinder
import com.github.vivchar.rendererrecyclerviewadapter.ViewModel
import com.github.vivchar.rendererrecyclerviewadapter.ViewRenderer
import kotlinx.coroutines.launch

class PayloadFragment : BaseFragment<FragmentListBinding>() {

	private val viewModel: PayloadViewModelVM by viewModels()
	private val adapter = MyAdapter()

	override fun createBinding(inflater: LayoutInflater, container: ViewGroup?) =
		FragmentListBinding.inflate(inflater, container, false)

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		super.onViewCreated(view, savedInstanceState)

		adapter.setDiffCallback(PayloadDiffCallback())
		adapter.registerRenderer(
			ViewRenderer<PayloadViewModel, ViewFinder>(
				R.layout.item_payload_square,
				PayloadViewModel::class.java
			) { model, finder, payloads: List<Any> ->
				finder.setOnClickListener(R.id.text) { viewModel.onItemClicked(model) }

				val textView = finder.find<TextView>(R.id.text)
				if (payloads.isEmpty()) {
					textView.text = model.text
					finder.setText(R.id.desciption, model.description)
				} else {
					val payload = payloads[0]
					if (payload == TEXT_CHANGED) {
						textView.rotation = 0f
						textView.animate().rotation(360f).start()
						textView.text = model.text
					} else if (payload == DESCRIPTION_CHANGED) {
						finder.setText(R.id.desciption, model.description)
					}
				}
			}
		)

		binding.recyclerView.adapter = adapter
		binding.recyclerView.layoutManager = GridLayoutManager(context, 3)
		binding.recyclerView.addItemDecoration(ItemOffsetDecoration(10))

		viewLifecycleOwner.lifecycleScope.launch {
			viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
				viewModel.state.collect { state ->
					adapter.setItems(state.items)
				}
			}
		}
	}

	private inner class PayloadDiffCallback : DefaultDiffCallback<PayloadViewModel>() {
		override fun areItemsTheSame(oldItem: PayloadViewModel, newItem: PayloadViewModel): Boolean {
			return oldItem.id == newItem.id
		}

		override fun getChangePayload(oldItem: PayloadViewModel, newItem: PayloadViewModel): Any? {
			return when {
				oldItem.text != newItem.text -> TEXT_CHANGED
				oldItem.description != newItem.description -> DESCRIPTION_CHANGED
				else -> super.getChangePayload(oldItem, newItem)
			}
		}
	}

	data class PayloadViewModel(val id: Int, val text: String, val description: String) : ViewModel

	companion object {
		const val TEXT_CHANGED = 1
		const val DESCRIPTION_CHANGED = 2
	}
}
